File: code/algo/getoffer/go/3/short_word_encode.go
package main

type WordEncodeTrie struct {
	child [26]*WordEncodeTrie
	depth int
}

func NewTrie() *WordEncodeTrie {
	return &WordEncodeTrie{
		child: [26]*WordEncodeTrie{},
		depth: 1,
	}
}

func (t *WordEncodeTrie) Insert(word string) {
	curr := t
	n := len(word)
	for i := n - 1; i >= 0; i-- {
		c := word[i] - 'a'
		if curr.child[c] == nil {
			curr.child[c] = NewTrie()
		}
		curr.child[c].depth = curr.depth + 1
		curr = curr.child[c]
	}
}

func minimumLengthEncoding(words []string) int {
	t := NewTrie()
	for _, w := range words {
		t.Insert(w)
	}
	var res int
	var dfs func(node *WordEncodeTrie)
	dfs = func(node *WordEncodeTrie) {
		var hasChild bool
		for i := 0; i < 26; i++ {
			if node.child[i] != nil {
				hasChild = true
				dfs(node.child[i])
			}
		}
		if !hasChild {
			res += node.depth
		}
	}
	dfs(t)
	return res
}

File: code/algo/getoffer/c++/3/short_word_encode.cpp
#include <algorithm>
#include <functional>
#include <string>
#include <vector>

class Trie {
public:
  Trie *child[26] = {nullptr};
  int depth = 1;

  Trie() = default;

  void Insert(const std::string &word) {
    Trie *curr = this;
    int n = word.size();
    for (int i = n - 1; i >= 0; --i) {
      int c = word[i] - 'a';
      if (curr->child[c] == nullptr) {
        curr->child[c] = new Trie();
      }
      curr->child[c]->depth = curr->depth + 1;
      curr = curr->child[c];
    }
  }
};

int minimumLengthEncoding(std::vector<std::string> &words) {
  Trie *root = new Trie();
  for (const std::string &word : words) {
    root->Insert(word);
  }

  int res = 0;
  std::function<void(Trie *)> dfs = [&](Trie *node) -> void {
    bool hasChild = false;
    for (int i = 0; i < 26; ++i) {
      if (node->child[i] != nullptr) {
        hasChild = true;
        dfs(node->child[i]);
      }
    }
    if (!hasChild) {
      res += node->depth;
    }
  };
  dfs(root);
  return res;
}
